### ISSUE TITLE
Double issue: paper idea and implementation

### ISSUE THREAD

**Issue by DiegoOrtego** [castorini/transformers-selective#1]
Hi,

Many thanks for sharing the paper code. I've been in the last weeks inspecting your work and related works in the literature and have 2 important doubts:

1) Related to the implementation: if I adapted and checked it well, it looks to me that the implementation of the regularization you propose leads to an overhead that makes training prohobitive with batch sizes like 128 or 256. This is because you have a double for loop to compute your regularization, e.g. in "modeling_selective_bert.py" you do:
regularizer = 0
for cc in correct_confidence:
    for wc in wrong_confidence:
        regularizer += torch.clamp(wc-cc, min=0) ** 2

Did you really use this implementation? Isn't this problematic? I've been running some experiments and is very slow.

2) Related to the paper content: I want to ask you why you didn't cite the ICML 2020 paper "[Confidence-aware learning for Deep Neural Networks](https://arxiv.org/abs/2007.01458)"? The proposed method is very similar, although the way of presenting the paper is different. I understand that the use case is different (NLP vs computer vision), but it looks to me that you took ideas from it.

Many thanks in advance.

**Comment by ji-xin** [comment]
Hi,

Thanks for your interest in our work.

1. You are correct that this is a double for-loop with a complexity of O(n^2), where n is the batch size. Assuming `len(wrong_confidence)`=m, then `len(correct_confidence)`=n-m. Therefore their product is at most n^2/4. Additionally, most of the time we have `len(wrong_confidence)` << n. Therefore the O(n^2) complexity comes with a pretty small constant. On the other hand, the transformer models are memory intensive, which already prohibits large batch sizes like 128. We did use the implementation for our experiments, and it worked well for the model and batch sizes we used. `reg-hist` is slower but that's because it does inference on the entire dataset frequently during the training process. The problem you mentioned indeed makes sense, and will require some additional approximation for the regularizer when large batch sizes are necessary.
2. Our paper and the paper by Moon et al. (2020) share some similarities, but as you mentioned, the domains, tasks, experiments, and presentation are different. As mentioned in Section 2 of our paper, the papers by Geifman and El-Yaniv (2017, 2019) are important related work of selective prediction and our paper borrowed ideas from the CV community. Both our paper and Moon et al. (2020) learned from these 2 papers, and we clearly cited Geifman and El-Yaniv (2017, 2019).

**Comment by DiegoOrtego** [comment]
Hi @ji-xin,

Many thanks for the clarifications. Much appreciated. Regarding 1, you may want to use the following to compute the regularization without using a loop (worrks in my use case with 256 batch size):
    confidence, pred_label = torch.softmax(prediction, dim=1).max(dim=1)
    correct = pred_label == labels
    incorrect = ~correct
    correct_confidence = torch.masked_select(confidence, correct)
    incorrect_confidence = torch.masked_select(confidence, incorrect)
    if torch.sum(correct)==0 or torch.sum(incorrect)==0:
        ranking_loss = torch.zeros(1, device=prediction.device)
    else:
        diff = incorrect_confidence.unsqueeze(1)[None,:,:]-correct_confidence.unsqueeze(1)[:,None,:]
        ranking_loss = torch.sum(torch.clamp(diff, min=0) ** 2)

**Comment by ji-xin** [comment]
Thanks for the suggestion! I'll update the code with this implementation when large batch sizes become necessary.
